cross_negative_script/cross_negative_script.gs

```diff
@@ -0,0 +1,216 @@
+ */
+function addAdGroupLevelNegatives() {
+  const campaignIterator = AdsApp.campaigns().withCondition("Status = ENABLED").get();
+  while (campaignIterator.hasNext()) {
+    const campaign = campaignIterator.next();
+    if (shouldProcessCampaign(campaign.getName())) {
+      log("Now processing the " + campaign.getName() + " campaign.");
+      log(" ");
+      const adGroups = [];
+      const adGroupIterator = campaign.adGroups().withCondition("Status = ENABLED").get();
+      while (adGroupIterator.hasNext()) {
+        const adGroup = adGroupIterator.next();
+        const keywords = getKeywords(adGroup);
+        adGroups.push({ adGroup, keywords });
+      }
+      adGroups.forEach(({ adGroup, keywords }) => {
+        if (keywords.length > 0) {
+          log(
+            "Adding " +
+              keywords.length +
+              " keywords from the " +
+              adGroup.getName() +
+              " ad group to the other ad groups in the " +
+              campaign.getName() +
+              " campaign."
+          );
+          log("");
+          adGroups.forEach(({ adGroup: otherAdGroup }) => {
+            if (adGroup.getId() !== otherAdGroup.getId()) {
+              addNegativesToAdGroup(otherAdGroup, keywords);
+            }
+          });
+          log("Adding complete.");
+          log(" ");
+        }
+      });
+    }
+  }
+}
+
+/**
+ * Determines if a campaign should be processed based on its name.
+ * @param {string} campaignName - The name of the campaign.
+ * @returns {boolean} - True if the campaign should be processed, false otherwise.
+ */
+function shouldProcessCampaign(campaignName) {
+  const includeRegex = INCLUDE_CAMPAIGN_NAMES.length > 0 ? new RegExp(INCLUDE_CAMPAIGN_NAMES.join("|"), "i") : null;
+  const excludeRegex = EXCLUDE_CAMPAIGN_NAMES.length > 0 ? new RegExp(EXCLUDE_CAMPAIGN_NAMES.join("|"), "i") : null;
+
+  const isIncluded = includeRegex ? includeRegex.test(campaignName) : true;
+  const isExcluded = excludeRegex ? excludeRegex.test(campaignName) : false;
+
+  return isIncluded && !isExcluded;
+}
+
+/**
+ * Retrieves keywords from an entity.
+ * @param {Object} entity - The entity from which to retrieve keywords.
+ * @returns {Array} - An array of keywords with their text and match type.
+ */
+function getKeywords(entity) {
+  const keywords = [];
+  const keywordIterator = entity.keywords().withCondition("Status = ENABLED").orderBy("Cost DESC").get();
+  while (keywordIterator.hasNext() && keywords.length < MAX_NEGATIVE_KEYWORDS) {
+    const keyword = keywordIterator.next();
+    keywords.push({
+      text: keyword.getText(),
+      matchType: keyword.getMatchType(),
+    });
+  }
+  return keywords;
+}
+
+/**
+ * Adds negative keywords to a campaign.
+ * @param {Object} campaign - The campaign to which negative keywords will be added.
+ * @param {Array} keywords - The keywords to add as negatives.
+ */
+function addNegativesToCampaign(campaign, keywords) {
```

Comment:
⚠️ Lines through 169 to 206 definitely aren't DRY. 

Make the function name more generic and the first argument name more generic and you'll only need one function.

As a general rule, repeating code that's identical or very similar is a bad move for lots of reasons (can share some resources if needed).

Reply:
100%. Didn't even realise I'd left those both in. Will update!